users/kuchosauronad0/rgblight_user.c

```diff
@@ -0,0 +1,157 @@
+#include "rgblight_user.h"
+extern rgblight_config_t rgblight_config;
+bool                     has_initialized;
+uint8_t                  base_hue;  // Hue value of base state
+uint8_t                  base_sat;  // Saturation value of base state
+uint8_t                  base_val;  // Brightness value of base state
+uint8_t                  base_mod;  // Animation mode of the base state
+
+void rgblight_savebase(void) {  // Save the current settings to ram
+  base_hue = rgblight_config.hue;
+  base_sat = rgblight_config.sat;
+  base_val = rgblight_config.val;
+  base_mod = rgblight_config.mode;
+}
+void rgblight_loadbase(void) {  // Load settings from ram to eeprom
+  rgblight_mode( base_mod );
+  rgblight_sethsv_eeprom_helper(base_hue, base_sat, base_val, true);
+}
+
+
+
+bool process_record_user_rgb(uint16_t keycode, keyrecord_t *record) {
+  if ((keycode >= QK_MOD_TAP && keycode <= QK_MOD_TAP_MAX) || (keycode >= QK_LAYER_TAP && keycode <= QK_LAYER_TAP_MAX)) {
+    keycode = keycode & 0xFF;
+  }
+  switch (keycode) {
+    case KC_RGB_T:  // Switch between underglow layer indication and normal qmk behaviour
+#if defined(RGBLIGHT_ENABLE) || defined(RGB_MATRIX_ENABLE)
+      if (record->event.pressed) {
+        userspace_config.rgb_layer_change ^= 1;
+        xprintf("rgblight layer change [EEPROM]: %u\n", userspace_config.rgb_layer_change);
+        eeconfig_update_user(userspace_config.raw);
+        if (userspace_config.rgb_layer_change) {
+          layer_state_set(layer_state);  // Immediately set the layer color (looks better)
+        }
+      }
+#endif  // RGBLIGHT_ENABLE
+      return false;
+      break;
+#ifdef RGBLIGHT_ENABLE
+    case RGB_MODE_FORWARD ... RGB_MODE_GRADIENT:  // quantum_keycodes.h L400 for definitions
+      if (record->event.pressed) {  // This disables layer indication, as it's assumed that if you're changing this ... you want that disabled
+        if (userspace_config.rgb_layer_change) {
+          userspace_config.rgb_layer_change = false;
+          xprintf("rgblight layer change [EEPROM]: %u\n", userspace_config.rgb_layer_change);
+          eeconfig_update_user(userspace_config.raw);
+        }
+      }
+      return true;
+      break;
+#endif  // RGBLIGHT_ENABLE
+  }
+  return true;
+}
+
+void keyboard_post_init_rgb(void) {
+# if defined(RGBLIGHT_ENABLE) || defined(RGBLIGHT_STARTUP_ANIMATION)
+  bool is_enabled = rgblight_config.enable;
+  if (userspace_config.rgb_layer_change) {
+    rgblight_enable_noeeprom();
+  }
+  if (rgblight_config.enable) {
+    layer_state_set_user(layer_state);
+    uint16_t old_hue = rgblight_config.hue;
+    rgblight_mode_noeeprom(RGBLIGHT_MODE_STATIC_LIGHT);
+    for (uint16_t i = 255; i > 0; i--) {
+      rgblight_sethsv_noeeprom((i + old_hue) % 255, 255, 255);
+      matrix_scan();
```

Comment:
Maybe I am missing context, but what does calling matrix_scan do here?

Reply:
I have matrix scan there, so that it will run the rgb matrix animations, rather than halt them.